src/services/organizeImports.ts

```diff
@@ -0,0 +1,235 @@
+/* @internal */
+namespace ts.OrganizeImports {
+    export function organizeImports(
+        sourceFile: SourceFile,
+        formatContext: formatting.FormatContext,
+        host: LanguageServiceHost) {
+
+        // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
+
+        // All of the old ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls = sourceFile.statements.filter(isImportDeclaration);
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const oldValidImportDecls = oldImportDecls.filter(importDecl => getExternalModuleName(importDecl.moduleSpecifier));
+        const oldInvalidImportDecls = oldImportDecls.filter(importDecl => !getExternalModuleName(importDecl.moduleSpecifier));
+
+        // All of the new ImportDeclarations in the file, in sorted order.
+        const newImportDecls = coalesceImports(sortImports(removeUnusedImports(oldValidImportDecls))).concat(oldInvalidImportDecls);
+
+        const changeTracker = textChanges.ChangeTracker.fromContext({ host, formatContext });
+
+        // Delete or replace the first import.
+        if (newImportDecls.length === 0) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[0]);
+        }
+        else {
+            // Note: Delete the surrounding trivia because it will have been retained in newImportDecls.
+            changeTracker.replaceNodeWithNodes(sourceFile, oldImportDecls[0], newImportDecls, {
+                useNonAdjustedStartPosition: false,
+                useNonAdjustedEndPosition: false,
+                suffix: getNewLineOrDefaultFromHost(host, formatContext.options),
+            });
+        }
+
+        // Delete any subsequent imports.
+        for (let i = 1; i < oldImportDecls.length; i++) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[i]);
+        }
+
+        return changeTracker.getChanges();
+    }
+
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    }
+
+    /* @internal */ // Internal for testing
+    export function sortImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        return stableSort(oldImports, (import1, import2) => {
+            const name1 = getExternalModuleName(import1.moduleSpecifier);
+            const name2 = getExternalModuleName(import2.moduleSpecifier);
+            Debug.assert(name1 !== undefined);
+            Debug.assert(name2 !== undefined);
+            return compareBooleans(isExternalModuleNameRelative(name1), isExternalModuleNameRelative(name2)) ||
+                compareStringsCaseSensitive(name1, name2);
+
+        });
+    }
+
+    function getExternalModuleName(specifier: Expression) {

```

Comment:
Could have a `ValidImportDeclaration` type that always has a valid specifier, and then not need to use this function or worry about `undefined` results.
```ts
interface ValidImportDeclaration extends ImportDeclaration { moduleSpecifier: StringLiteral; }
function isValidImportDeclaration(node: Node): node is ValidImportDeclaration {
    return isImportDeclaration(node) && isStringLiteral(node.moduleSpecifier);
}
```
then:
```ts
const oldImportDecls = sourceFile.statements.filter(isValidImportDeclaration);
```
(We could technically allow `NoSubstitutionTemplateLiteral`s here, but they're actually compile errors so not really valid.)